File: xg-frontend/apps/web/src/api/fieldCatalog.ts
import api from './index';

/**
 * 后端 yaml 字段目录的前端镜像。一份 yaml 驱动:后端 SqlBuilder + AI tool schema +
 * 前端 chip 渲染。前端拿到这份描述后,知道每个 filter 应该用什么样的控件、选项从哪来、
 * 父级是谁、是否被租户开关门控。
 *
 * 加新字段 = 改 yaml 一处,这里类型不需要动。
 */
export interface FieldCatalogResponse {
  page: string;
  fields: FieldDef[];
}

export interface FieldDef {
  key: string;
  label: string;
  /** AI prompt 用的描述,前端目前不展示。 */
  description?: string | null;
  type: 'text' | 'enum';
  /** 父级 filter 的 key,用于级联(选了父级才显示子级)。 */
  parent?: string | null;
  /** 租户开关名:tenantSettings 上对应字段为 true 才渲染。null 表示无门控。 */
  tenant_setting?: string | null;
  /** 静态枚举选项 (e.g. gender / status / 学院列表)。 */
  options?: { value: string; label: string }[] | null;
  /** 父级值分组的静态选项 (e.g. 学院↔专业) — parent 选了某 college,子选项就是对应 group。 */
  cascading_static?: CascadingStatic | null;
  /** 动态选项:调一个 url 拿,父级值通过 scoped_by 传参。 */
  dynamic?: DynamicHint | null;
  /** AI 别名,前端不直接用,只在 LLM prompt 里。 */
  aliases?: { phrase: string; value: string }[] | null;
}

export interface CascadingStatic {
  parent_key: string;
  groups: Record<string, string[]>;
}

export interface DynamicHint {
  url?: string | null;
  key?: string | null;
  /** 父级 filter 的 key 列表,作为请求参数透传 */
  scoped_by?: string[] | null;
}

export function getFieldCatalog(page: string): Promise<FieldCatalogResponse> {
  return api.get(`/field-catalog/${page}`).then((res) => res.data);
}

File: xg-backend/xg-business/src/main/java/com/xg/business/fieldcatalog/model/FieldCatalog.java
package com.xg.business.fieldcatalog.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

/**
 * 字段目录:一份 yaml 描述一个页面能按什么筛 + 每条筛选项怎么取数 / 怎么显示 / AI 怎么映射别名。
 *
 * 一个 page 对应一个 yaml 文件 (e.g. field-catalog/student.yaml),启动时由
 * {@link com.xg.business.fieldcatalog.loader.FieldCatalogLoader} 全部读入内存,
 * 由 {@link com.xg.business.fieldcatalog.service.FieldCatalogService} 暴露查询。
 *
 * 设计为不可变 record,加载后不允许修改 (热更新 P0 不做)。
 */
public record FieldCatalog(String page, List<FieldDef> fields) {

    public record FieldDef(
            String key,
            String label,
            /** AI prompt 用的字段说明,可选;不填则只用 label,LLM 拿到的提示就比较干。 */
            String description,
            String type,
            SqlConfig sql,
            OptionsConfig options,
            String parent,
            Visibility visibility,
            List<Alias> aliases
    ) {}

    public record SqlConfig(
            String strategy,
            String column,
            List<String> columns,
            String residentialType,
            String jsonbPath
    ) {}

    public record OptionsConfig(
            /* yaml key 是 'static',Java 关键字冲突,这里用 staticOptions */
            @JsonProperty("static") List<EnumOption> staticOptions,
            CascadingStatic cascadingStatic,
            DynamicSource dynamic
    ) {}

    /**
     * 父级值分组的静态选项。例如 college=计算机学院 → ["软件工程", ...]。
     * 父级未选时前端展开所有 group 的扁平合集。
     */
    public record CascadingStatic(
            /** 父级 filter 的 key,通常和 FieldDef.parent 相同;冗余写一份是为了 yaml 单点可读。 */
            String parentKey,
            /** 父级值 → 该子级可选项 (option label 直接用 value)。 */
            java.util.Map<String, List<String>> groups
    ) {}

    public record EnumOption(String value, String label) {}

    public record DynamicSource(
            String source,
            String column,
            String url,
            String key,
            List<String> scopedBy
    ) {}

    public record Visibility(String tenantSetting) {}

    public record Alias(String phrase, String value) {}
}
